### ISSUE TITLE
Adding particles to SimArray

### ISSUE THREAD

**Issue by stinsong4100** [pynbody/pynbody#299]
Would it be possible to add particles to SimArrays?

SimArrays are an extremely useful data structure like numpy arrays.  However, it is currently not possible to expand SimArrays.  Could it be made possible to add elements to SimArrays like np.append or np.insert?

This is echoing SunWon Kwak's request on the pynbody-users group who wanted to 2 galaxies together.

**Comment by apontzen** [comment]
They just wrap numpy arrays so anything you can do with a numpy array, you can do with a SimArray...

**Comment by stinsong4100** [comment]
Andrew said on pynbody-users

> You can use numpy.insert, but it returns a new array rather than altering an existing one.
>
> For pynbody itself, the only supported thing you can do is create a new empty file, e.g.
>
> f = pynbody.new(gas=len(f1.gas)+len(f2.gas), star=len(f1.star)+len(f2.star), dm=len(f1.dm)+len(f2.dm))
> 
> then manually copy in the arrays.
>
> It wouldn't be too difficult to write a pynbody.new_from(f1,f2) function that does this automatically, if it 
> would be useful

I see.  I posed the question incorrectly.  Let me try again:

I'd ideally like to use SimSnaps for a chemical evolution code and run it in python.  Each timestep, I'd like to add particles.  Is there a good way to make that happen?

What I love about SimSnaps are the decorators.  They make creating arrays like 'OFe' or 'FeH' really easy.  Here's my code using numpy arrays.

```
        new_star = [time,mass,mass,self.Z]
        new_star.extend(self.abunds.values())
        self.stars = np.append(self.stars,np.array([tuple(new_star)],
                                                   dtype=star_type))
```

Right now, with SimArrays, if I try
`s.s['pos'] = np.append(s.s['pos'],[[1,2,4]],axis=0)`
I get
`ValueError: could not broadcast input array from shape (101,3) into shape (100,3)`
That seems like it is behaving differently than numpy arrays.

It sounds like you're pynbody.new_from function would do what I want.  
It would be even better if it were possible to do it inline
`s.s.extend(new_stars)`

I guess this is impossible since the underlying structures are numpy arrays which only allow creating longer copies.

Thanks for your input!

**Comment by apontzen** [comment]
You could take the C++ STL vector approach, where you allocate an array that's longer than you need but only view a bit of it. This is easy with the numpy view mechanism...

There's no in-built support for this in terms of allocating extra memory for an entire SimSnap, though.

**Comment by apontzen** [comment]
BTW the problem with

```
s.s['pos'] = np.append(s.s['pos'],[[1,2,4]],axis=0)
```

is that the SimSnap enforces the correct length at present, so I think the error is generated where you copy it back into the SimSnap. If your main requirement is to use SimArrays, you could always just store them in a dictionary.

**Comment by stinsong4100** [comment]
Allocating at the beginning does sound like the most sensible option.  Thanks!

Do you want to leave this open until you implement pynbody.new_from(s1,s2)?

**Comment by apontzen** [comment]
I don't have time to implement this right now, so I'll only leave it open if you're going to?
